### PR TITLE
Development vdc payment

### PR DESCRIPTION
**Pull request by m-motawea**

### CODE REVIEW

File: jumpscale/packages/vdc/chats/extend_k8s.py [hunk 1]
```diff
@@ -41,17 +45,35 @@ def add_node(self):
...
 
         self.md_show_update("Payment successful")
+        initialization_wallet_name = j.core.config.get("VDC_INITIALIZATION_WALLET")
+        old_wallet = deployer._set_wallet(initialization_wallet_name)
         wids = deployer.add_k8s_nodes(self.node_flavor, public_ip=self.public_ip)
         if not wids:
-            # TODO: what to do in case a deployment failed!
+            j.sals.billing.issue_refund(payment_id)
             self.stop("failed to add nodes to your cluster. please contact support")
+        self.md_show_update("processing tansaction...")
```
Comment (waleedhammam) [inline jumpscale/packages/vdc/chats/extend_k8s.py:59]:
```suggestion
        self.md_show_update("Processing transaction...")
```

File: jumpscale/sals/vdc/size.py [hunk 2]
```diff
@@ -4,19 +4,21 @@
...
-S3_AUTO_TOPUP_FARMS = ["lochristi_dev_lab", "lochristi_dev_lab"]
-ZDB_FARMS = ["lochristi_dev_lab", "lochristi_dev_lab"]
-PREFERED_FARM = "lochristi_dev_lab"
-NETWORK_FARM = "lochristi_dev_lab"
+INITIAL_RESERVATION_DURATION = 1  # in hours
+
+S3_AUTO_TOPUP_FARMS = ["freefarm", "freefarm"]
+ZDB_FARMS = ["freefarm", "freefarm"]
+PREFERED_FARM = "freefarm"
+NETWORK_FARM = "freefarm"
```
Comment (waleedhammam) [inline jumpscale/sals/vdc/size.py:21]: what about devnet and mainnet

**Review by ranatrk — APPROVED** [review]

File: jumpscale/sals/vdc/deployer.py [hunk 3]
```diff
@@ -344,21 +361,104 @@ def deploy_vdc_kubernetes(self, farm_name, scheduler, cluster_secret):
...
         """deploys a new vdc
         Args:
             cluster_secret: secretr for k8s cluster. used to join nodes in the cluster (will be stored in woprkload metadata)
             minio_ak: access key for minio
             minio_sk: secret key for minio
             farm_name: where to initialize the vdc
         """
+        if not self.check_capacity(farm_name):
+            raise j.exceptions.Validation(
+                f"not enough resources in farm {farm_name} to deploy vdc of flavor {self.flavor}"
```
Comment (xmonader) [inline jumpscale/sals/vdc/deployer.py:459]: can we inline the flavor data here?

File: jumpscale/sals/vdc/deployer.py [hunk 4]
```diff
@@ -470,15 +570,26 @@ def deploy_vdc(self, minio_ak, minio_sk, farm_name=PREFERED_FARM):
                 self.monitoring.deploy_stack()
             except j.exceptions.Runtime as e:
                 # TODO: rollback
-                self.error(f"failed to deploy monitoring stack on VDC cluster due to error {str(e)}")
+                self.error(f"failed to deploy monitoring stack on vdc cluster due to error {str(e)}")
```
Comment (xmonader) [inline jumpscale/sals/vdc/deployer.py:573]: VDC should be uppercase

File: jumpscale/sals/vdc/kubernetes.py [hunk 5]
```diff
@@ -330,7 +331,7 @@ def _add_workers(
                 except DeploymentFailed:
                     if public_wids[idx]:
                         self.zos.workloads.decomission(public_wids[idx])
-                    self.vdc_deployer.error(f"failed t.o deploy kubernetes worker wid: {wid}")
+                    self.vdc_deployer.error(f"failed to deploy kubernetes worker wid: {wid}")
                     pass
```
Comment (xmonader) [inline jumpscale/sals/vdc/kubernetes.py:335]: no need for the pass here?

**Review by xmonader — APPROVED** [review]

File: jumpscale/packages/vdc/chats/new_vdc.py [hunk 6]
```diff
@@ -55,32 +62,61 @@ def deploy(self):
...
             self.deployer = self.vdc.get_deployer(password=self.vdc_secret.value, bot=self)
         except Exception as e:
             j.logger.error(f"failed to initialize VDC deployer due to error {str(e)}")
-            self.vdc.refund_payment(trans_hash)
+            j.sals.billing.issue_refund(payment_id)
             j.sals.vdc.delete(self.vdc.vdc_name)
             self.stop("failed to initialize VDC deployer. please contact support")
 
         self.md_show_update("Deploying your VDC...")
+        initialization_wallet_name = j.core.config.get("VDC_INITIALIZATION_WALLET")
```
[the remark this answers is not in the document]
Comment (m-motawea) [reply jumpscale/packages/vdc/chats/new_vdc.py:87]: it is not automatically created. it up to the user to configure it

**Pull request merged**